nixos/modules/tasks/swraid.nix

```diff
@@ -1,17 +1,50 @@
-{ pkgs, ... }:
+{ pkgs, lib, config, ... }:
 
+with lib;
+let cfg = config.services.mdadm;
+in
 {
 
-  environment.systemPackages = [ pkgs.mdadm ];
+  options = {
 
-  services.udev.packages = [ pkgs.mdadm ];
+    services.mdadm = {
 
-  systemd.packages = [ pkgs.mdadm ];
+      monitor = {
+        enable = mkEnableOption "Enable the mdadm monitoring daemon";
+      };
 
-  boot.initrd.availableKernelModules = [ "md_mod" "raid0" "raid1" "raid10" "raid456" ];
+      adminAddr = mkOption {
+        type = types.str;
+        example = "admin@example.org";
+        description = "E-mail address of the server administrator.";
+      };
+    };
+  };
 
-  boot.initrd.extraUdevRulesCommands = ''
-    cp -v ${pkgs.mdadm}/lib/udev/rules.d/*.rules $out/
-  '';
+  config = mkMerge [
+    {
+      environment.systemPackages = [ pkgs.mdadm ];
+
+      services.udev.packages = [ pkgs.mdadm ];
+
+      systemd.packages = [ pkgs.mdadm ];
+
+      boot.initrd.availableKernelModules = [ "md_mod" "raid0" "raid1" "raid10" "raid456" ];
+
+      boot.initrd.extraUdevRulesCommands = ''
+        cp -v ${pkgs.mdadm}/lib/udev/rules.d/*.rules $out/
+      '';
+    }
+
+    (mkIf cfg.monitor.enable
+        {
+          systemd.services.mdadm-monitor = {
+            description = "Monitor RAID disks";
+            wantedBy = [ "multi-user.target" ];
+            script = "${pkgs.mdadm}/bin/mdadm --monitor -m ${assert cfg.adminAddr != null; cfg.adminAddr} --scan";
+          };
```

Comment:
That is a possibility, yes. I was under the (probably wrong) impression that the original change that introduced the units didn't really do anything helpful. I'd be happy to do it this way as well. I'm planning to consider adding more tests for the things that we expect to happen so we don't have to scratch our heads in the future. This may likely need a few days until I get around to adding more effort here, but I'll get to it. :)

Reply:
It'd be awesome if you could add tests indeed! That said if you don't have the time, I think this PR already makes things better so could probably land :)